### ISSUE TITLE
Firebase issue - UpdateValueAsync and listeners

### ISSUE THREAD

**Issue by craNich** [firebase/quickstart-unity#432]
Unity editor version: 
Firebase Unity SDK version: 6.2.2
Firebase plugins in use (Auth, Database, etc.): Auth, Database
Additional SDKs you are using (Facebook, AdMob, etc.):  Mapbox
Platform you are using the Unity editor on (Mac, Windows, or Linux): Windows
Platform you are targeting (iOS, Android, and/or desktop): Android

Hello. I just installed new version of sdk and it still didnt fix the Update issue for me.
I have an issue with update and listener. I am getting an error while I am updating two different paths with listener attached to one of this path. When I make two updates with each updating just one path, everything works fine.
The error is actualy on the listener:
System.NullReferenceException: Object reference not set to an instance of an object.
Edit: Read bellow - Edit. 

Also this update actualy gave me another issue with listener. After the update of firebase one of the listeners I have in my app is called twice doing same operation twice in same time which leads to duplication of some things. Very very strange. Changed to previous version with same code = no issues with that. Of course I can fix it with some checking if its been already processed, but ...
I think there is something wrong with listeners overall. No update included here. Just Pushing a new record, then listening for an event - called twice for no reason I think. At least in previous version it was called just once.

Edit:
Tried to debug some things, there is an example of which works and which doesnt. Little bit differenet example of the bug, but I think it has something to do with it. 

Works:
Dictionary<string, object> childUpdates = new Dictionary<string, object>();
childUpdates["node3/" + someString] = true;
firebaseRootRef.Child("node1").Child(node2).UpdateChildrenAsync(childUpdates);

Doesnt work (Object reference not set to an instance of an object on the listener Handle function):
Dictionary<string, object> childUpdates = new Dictionary<string, object>();
childUpdates["node2/node3/" + someString] = true;
firebaseRootRef.Child("node1").UpdateChildrenAsync(childUpdates);

- The listener with error is listening to path: node1/node2/node4, when it reads the changes and do something. The error I get is when reading the snapshot. Actualy when using the "path/path" update it does trigger the listener in node4, when i am updating only the node3, which has same parrent node. I dont think this is intended. So I think it gets some wrong snapshot and then error occures.
Also the error doesnt occur, when I first update the data in node4 and then update the data in node3. But its realy very hard to debug.

It seems like, the last child in args.snapshot has much more children than it should has. Its very weird, I have same structure with just different ids, but the snapshot children count is in the last child different from others. When I look into firebase, there is nothing wrong with it. Everything looks same.

This is the debug log from unity, which shows children count of childs in snasphot, which is recieved by listener in response to update change of some parrent node. (which shouldnt be even invoked i believe, since it doesnt do any update on the node I listen to). And actualy in the last child which has weird count of children, there is no child called "duration" found. And it actualy has child of its key. Am I missing something here?
So it returns old childs and new child which has inside the other old childs and itself? I just dont get it. 

![snapshot](https://user-images.githubusercontent.com/53269312/62130219-d3f15780-b2d8-11e9-9445-1a96d9a53dbe.png)

But of course, when I dont use update with the "/path/path" (I cant of course update multiple paths without that syntax) it just returns what I expect to. Same children count for every child.

Edit2: Okey, I looped through the snapshot recieved from listener. It contains all keys with duration and spawntime, except the last one which contains spawntime and all keys with duration/spawntime.

**Comment by alexames** [comment]
Hi @craNich, I'm looking into what might be happening here. 

I want to make sure I understand that case where you're seeing the error so that I can reliably recreate the situation you're talking about.

Would this code recreate the issue you're seeing? I'm going to test it in a bit, but I want to make sure I have the test case set up right. 
```cs
DatabaseReference testRef = FirebaseDatabase.DefaultInstance.GetReference("Test");
// Should only react to changes at Test/Base/Child/ChildWithListener and below.
testRef.Child("Base").Child("Child").Child("ChildWithListener").ValueChanged += 
    (object sender, ValueChangedEventArgs e) => {
      if (e.DatabaseError != null) {
        Debug.LogError(e.DatabaseError.Message);
        return;
      }
      Debug.Log("Received value changed event at " + e.Snapshot.Key);
      if (e.Snapshot.Key != "ChildWithListener") {
        Debug.LogError("Expected key = \"ChildWithListener\", got \"" +  + e.Snapshot.Key + "\"");
      }
      Debug.Log("Snapshot JSON is: " + e.Snapshot.GetRawJsonValue());
    };

// Create an update to set the relative path Child/ChildWithoutListener to true
// at the location Test/Base.
Dictionary<string, object> childUpdates = new Dictionary<string, object>();
childUpdates["Child/ChildWithoutListener"] = true;

// Apply the update. Only Test/Base/Child/ChildWithoutListener should be updated.
testRef.Child("Base").UpdateChildrenAsync(childUpdates);

// After applying the update, the ValueChanged callback is called when it should
// not have been, since Test/Base/Child/ChildWithListener was not modified.
```

**Comment by craNich** [comment]
@alexames 
Hello. I am not 100% sure this will make the error. But the callback will be called at least I think.

Error will occure if the "ChildWithListener" has below some children also. So the snapshot returns these children. Then going through these children with 
foreach (var child in args.Snapshot.Children) { Debug.Log(child.ChildrenCount); }
will return the error. Or well, not an error in this example, but the last child will return something wrong.

I can provide my code if necessary to an email.

**Comment by alexames** [comment]
Yeah, it would be very helpful if you could post a minimal example that reproduces the error. It doesn't have to be a full project, just the bit that is important to reproducing the error, with what outcome is occurring for you and what outcome is expected.

**Comment by craNich** [comment]
Can I send it to your email?

**Comment by alexames** [comment]
You can, but if possible it would be preferable to just post the relevant snippet here so that others can see what the issue is too. Remember I don't need a whole project, just the minimal number of lines that reproduce the issue.

**Comment by craNich** [comment]
https://pastebin.com/RWx8Y6Vs

Part of firebase database;
Btw. I have listener on the "coins_info" too. But not like the listener in "coins_expiration" this one is never called when I update "coins_smoked" node. 

![firebase_error](https://user-images.githubusercontent.com/53269312/62170872-3ecc7e00-b32d-11e9-9bf7-8e6af2a159bc.png)

**Comment by craNich** [comment]
This is like the snapshot data I get from listener from what I have been seeing so far
![snapshotdata](https://user-images.githubusercontent.com/53269312/62190743-642fab00-b372-11e9-824a-00c473ee58cc.png)

**Comment by alexames** [comment]
Thanks for the code, I'll let you know what we find.

**Comment by craNich** [comment]
Thanks!

**Comment by craNich** [comment]
Just reporting: This bug is fixed for me in newest firebase unity sdk version.

**Comment by stewartmiles** [comment]
@alexames it this expected?

**Comment by a-maurice** [comment]
Yes, this should have been fixed with the 6.5.0 release.  Sorry for not keeping this issue up to date, and thank you for letting us know that this issue has now been solved for you!
